libparsec/crates/client/src/certif/user_update_profile.rs

```diff
@@ -0,0 +1,179 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use libparsec_client_connection::ConnectionError;
+use libparsec_protocol::authenticated_cmds;
+use libparsec_types::prelude::*;
+
+use super::{
+    greater_timestamp, store::CertifStoreError, CertificateBasedActionOutcome, CertificateOps,
+    GreaterTimestampOffset, InvalidCertificateError, InvalidKeysBundleError,
+};
+use crate::EventTooMuchDriftWithServerClock;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertifUpdateUserProfileError {
+    #[error("Component has stopped")]
+    Stopped,
+    #[error("Cannot reach the server")]
+    Offline,
+    #[error("Cannot revoke oneself")]
+    UserIsSelf,
+    #[error("User not found")]
+    UserNotFound,
+    #[error("Author not allowed")]
+    AuthorNotAllowed,
+    #[error("User revoked")]
+    UserRevoked,
+    #[error("Our clock ({client_timestamp}) and the server's one ({server_timestamp}) are too far apart")]
+    TimestampOutOfBallpark {
+        server_timestamp: DateTime,
+        client_timestamp: DateTime,
+        ballpark_client_early_offset: f64,
+        ballpark_client_late_offset: f64,
+    },
+    #[error("There is no key available in this realm for encryption")]
+    NoKey,
+    #[error(transparent)]
+    InvalidKeysBundle(#[from] Box<InvalidKeysBundleError>),
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ConnectionError> for CertifUpdateUserProfileError {
+    fn from(value: ConnectionError) -> Self {
+        match value {
+            ConnectionError::NoResponse(_) => Self::Offline,
+            // TODO: handle organization expired and user revoked here ?
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+
+impl From<CertifStoreError> for CertifUpdateUserProfileError {
+    fn from(value: CertifStoreError) -> Self {
+        match value {
+            CertifStoreError::Stopped => Self::Stopped,
+            CertifStoreError::Internal(err) => err.into(),
+        }
+    }
+}
+
+pub(super) async fn update_profile(
+    ops: &CertificateOps,
+    user: UserID,
+    new_profile: UserProfile,
+) -> Result<CertificateBasedActionOutcome, CertifUpdateUserProfileError> {
+    if ops.device.user_id == user {
+        // TODO: should an admin be able to downgrade themselves
+        return Err(CertifUpdateUserProfileError::UserIsSelf);
```

Comment:
This TODO needs to be addressed before merging the PR --> please update the PR description to add it as an open question.

Also, tests will need to be added either to accept downgrading or test that is not possible.

Reply:
You should have a look at the certificate ops validation code (look for `check_user_update_certificate_consistency` function)

tl;dr: the code contains `// 2) Check the certificate is not self-signed` ;-)